### PR TITLE
Segregated Witness Support - WIP

### PR DESCRIPTION
**Pull request by rubensayshi**
# WORK IN PROGRESS

getting shit to work before making it pretty; almost done with getting shit to work!

depends on: https://github.com/bitcoinjs/bitcoinjs-lib/pull/540

first time spending a p2wpkh (6a539927494f3d6f07078dd1c70a28e0c4910662593072305529eaa66257f59d) and p2wpkh as p2sh (4704de796b8ea819f262eb3399fad85b838327497639da2ceafbd04a08a06be3) with bitcoinjs-lib :D
### ToDo
- [x] Output script creation for P2WPKH
- [x] Output script creation for P2WSH
- [x] Transaction encoding / decode
- [x] SignatureHash v2
- [x] Transaction signing of P2WPKH
- [x] Transaction signing of P2WPKH as P2SH
- [x] Transaction signing of P2WSH
- [x] `toBuffer` should be able to return both segwit and non-segwit TX
  - [x] fix `getTxId` using the segwit TX atm
- [x] BIP142 addresses for P2WPKH
- [ ] Add `value` arg to `addInput` and use it in `TransactionBuilder.sign` instead of `value` arg there?
- [x] Refactoring of changes made so far
- [ ] More tests
  - [x] Add tests for mix of non witness and witness (already manually tested and works)
  - [ ] Integration tests (and an API to already run them against segnet)
  - [ ] Check bitcoin core projects for any vectors that we can copy
- [ ] Review and discuss changes for BIP142 addresses
- [ ] Review and discuss changes to `TransactionBuilder.sign`
- [ ] Review and discuss changes to `TransactionBuilder.extractInput` / `script.classifyInput`
- [ ] Cache `hashPrevouts`, `hashSequence` and `hashOutputs` for all signatures in a transaction (since that was one of the main reasons for changing the signatureHash xD)
### Usage

**Signing Native P2WPKH**

``` js
transactionBuilder.sign(0, keyPair, null, null, /* segWit = */ true, /* txIn.value = */ parseInt(0.9998 * 1e8)))
```

**Signing P2WPKH nested in P2SH**

``` js
var pubKeyHash = bcrypto.hash160(keyPair.getPublicKeyBuffer())
var redeemScript = bscript.segWitPubKeyHashOutput(pubKeyHash)

transactionBuilder.sign(0, keyPair, redeemScript, null, /* segWit = */ true, /* txIn.value = */ parseInt(0.9998 * 1e8)))
```
### References
- BIP141 (general segwit) https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki
- BIP142 (segwit address) https://github.com/bitcoin/bips/blob/master/bip-0142.mediawiki
- BIP143 (signature verification with new signature hash) https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
- BIP144 (peer services) https://github.com/bitcoin/bips/blob/master/bip-0144.mediawiki
- Staging branch https://github.com/sipa/bitcoin/tree/segwit

### CODE REVIEW

File: src/transaction.js [hunk 1]
```diff
@@ -170,9 +223,100 @@ var VALUE_UINT64_MAX = new Buffer('ffffffffffffffff', 'hex')
...
+
+  if (segWit) {
+    return this.hashForSignatureV2(inIndex, prevOutScript, amount, hashType)
+  } else {
+    return this.hashForSignatureV1(inIndex, prevOutScript, hashType)
+  }
+}
+
+Transaction.prototype.hashForSignatureV2 = function (inIndex, prevOutScript, amount, hashType) {
+  var hashPrevouts = new Buffer("00" * 32, 'hex')
```
Comment (fanatid) [inline src/transaction.js]: @rubensayshi it isn't python :laughing:

Comment (rubensayshi) [inline src/transaction.js]: hehe yea noticed that yday, been switching between projects too much :P

**Comment by dcousens** [comment]
@rubensayshi when is the next time you can be available on slack?   It'd be good to talk over how this [and future changes] will integrate with `TransactionBuilder` and `Transaction`.

**Comment by rubensayshi** [comment]
@dcousens I've got slack on my phone so you can always poke me but I usually prefer keeping my weekends code-free, though this weekend is an exception and working on this PR atm :D
poke me whenever you have time this week, I'm available all week (GMT+1)

File: src/types.js [hunk 2]
```diff
@@ -13,6 +13,7 @@ function Buffer256bit (value) { return nBuffer(value, 32) }
 var UINT53_MAX = Math.pow(2, 53) - 1
 function UInt2 (value) { return (value & 3) === value }
 function UInt8 (value) { return (value & 0xff) === value }
+function UInt16 (value) { return (value >> 0) === value } // @TODO: is this correct? copy paste of UInt32
```
Comment (rubensayshi) [inline src/types.js]: is this okay to do ?

File: src/address.js [hunk 3]
```diff
@@ -38,8 +63,15 @@ function toOutputScript (address, network) {
   network = network || networks.bitcoin
 
   var decode = fromBase58Check(address)
+
   if (decode.version === network.pubKeyHash) return bscript.pubKeyHashOutput(decode.hash)
   if (decode.version === network.scriptHash) return bscript.scriptHashOutput(decode.hash)
+  if (decode.version === network.segWitPubKeyHash) {
+    if (decode.segWitVersion !== 0 && decode.segWitPadding !== 0) {
+      throw new TypeError(address + ' is too long for normal address and not a valid segWit address')
```
Comment (rubensayshi) [inline src/address.js]: not sure if this should go here or in `fromBase58Check`, considering the length checks are in `fromBase58Check`, this code has gotten a bit akward now ...

consider one of the test vectors for `fromBase58Check` we had of an address with hash length of 22, it no longer throws a 'is too long' exception and because we're unaware of `network` in `fromBase58Check` we can't really check if it's sane either ...

what was the original reason for seperating `fromBase58Check` and putting the exceptions there?

File: src/transaction_builder.js [hunk 4]
```diff
@@ -401,6 +447,18 @@ TransactionBuilder.prototype.sign = function (index, keyPair, redeemScript, hash
 
           break
 
+        case 'segwitpubkeyhash':
```
Comment (rubensayshi) [inline src/transaction_builder.js]: for P2WPKH as P2SH we place the `redeemScript` (short version) in the `scriptSig` but we'll use the long version when generating the signatureHash (hence we added `signatureScript`)

File: src/transaction.js [hunk 5]
```diff
@@ -250,8 +399,22 @@ Transaction.prototype.getId = function () {
   return [].reverse.call(this.getHash()).toString('hex')
 }
 
-Transaction.prototype.toBuffer = function () {
-  var buffer = new Buffer(this.byteLength())
+Transaction.prototype.toBuffer = function (withWitness) {
```
Comment (rubensayshi) [inline src/transaction.js]: I think it would make more sense that `withWitness` would default to TRUE, because the only reason for wanting the non-witness is to `getId` or if you want to submit it to an old node or some old piece of software

and in the tests that would mean I'll add a `txHexWithoutWitness` instead of the `txHexWithWitness` that I've added so far in a few places ... 

/discuss

**Comment by rubensayshi** [comment]
should I update other places where the `BufferWriter` can be used too?

File: src/transaction.js [hunk 6]
```diff
@@ -250,8 +399,22 @@ Transaction.prototype.getId = function () {
...
+  // console.log('tx.Buffer')
+
+  // @TODO: it would be nicer if the marker/flag would be set when witness is set on one of the txIn...
```
Comment (rubensayshi) [inline src/transaction.js]: /discuss

**Comment by dcousens** [comment]
Not yet,  lets keep the changes as atomic as necessary.  We can refactor later.

**Comment by dcousens** [comment]
@greenaddress lots of WIP here,  be weary of using this until its ready. :+1:

**Comment by greenaddress** [comment]
@dcousens ack thanks!

**Comment by bityogi** [comment]
ACK. Thanks for working on this.

**Comment by rubensayshi** [comment]
depends on https://github.com/bitcoinjs/bitcoinjs-lib/pull/524

File: .gitignore [hunk 7]
```diff
@@ -1,3 +1,5 @@
 bitcoin.js
 coverage
 node_modules
+
+_dev.*.js
```
Comment (rubensayshi) [inline .gitignore]: hope you don't mind, but this makes it easy to keep debug files around

**Comment by rubensayshi** [comment]
had to do a refactor of `extractInputs` and `__build` to be able to support P2SH[ P2WSH[ multsig / p2pkh / p2wpkh ] ]

I'm gonna see if I can seperate those changes into a seperate PR to make it easier to review (doing that right now)

created an overview of different scripts that are supported and what is what; https://gist.github.com/rubensayshi/fbe39657e90ef416d411

**Comment by dcousens** [comment]
@rubensayshi could we start writing up the CHANGELOG for this such that the changes can be quicker to review?

**Comment by dcousens** [comment]
Might move this to `2.4.0`

**Comment by bpdavenport** [comment]
Curious if anyone is still working on this. @dcousens is there any other work on segwit happening besides this pull request?

**Comment by dcousens** [comment]
@bpdavenport yes,  but it isn't public at this time.  I intend to have SegWit support before activation :+1:

**Comment by rubensayshi** [comment]
@dcousens I was about to ask you about this, because I noticed the refactoring of a few things ...

let me know if there's anything I can do!

**Comment by btcdrak** [comment]
It would be great to get this finished and merged so it can be used on testnet. Segwit is merged to bitcoin/master and active on Bitcoin testnet. It means people can start working on integration in earnest.

**Comment by dcousens** [comment]
@btcdrak time permitting, I'll see if I can't work towards a 10th July PR date.

File: src/transaction.js (hunk 5 above)
Comment (dcousens) [inline src/transaction.js]: I don't know if this is the right way to do it,  I've been playing with `toWitnessBuffer`,  or `WitnessTransaction` inheriting from `Transaction` where `toBuffer` is overwritten... still not sure.
I think we might just be better off detecting if a witness is needed and not bothering with a flag.

File: src/transaction.js [hunk 8]
```diff
@@ -170,9 +229,100 @@ var VALUE_UINT64_MAX = new Buffer('ffffffffffffffff', 'hex')
...
+  typeforce(types.tuple(types.UInt32, types.Buffer, /* types.UInt8 */ types.Number, types.maybe(types.Boolean)), arguments)
+
+  if (segWit) {
+    return this.hashForSignatureV2(inIndex, prevOutScript, amount, hashType)
+  } else {
+    return this.hashForSignatureV1(inIndex, prevOutScript, hashType)
+  }
+}
+
+Transaction.prototype.hashForSignatureV2 = function (inIndex, prevOutScript, amount, hashType) {
```
Comment (dcousens) [inline src/transaction.js]: Why is this necessary?

File: src/transaction.js [hunk 9]
```diff
@@ -4,9 +4,15 @@ var bufferutils = require('./bufferutils')
...
 var types = require('./types')
+var BufferWriter = require('./bufferwriter')
+
+var ADVANCED_TRANSACTION_MARKER = 0
+var ADVANCED_TRANSACTION_FLAG = 1
 
 function Transaction () {
   this.version = 1
+  this.marker = null
+  this.flag = null
```
Comment (dcousens) [inline src/transaction.js]: Are these ever going to change?
My understanding was the `0` marker was to allow for easy detection in serialization?

**Comment by dcousens** [comment]
Lots of debugging `console.log` lefts in here

File: .gitignore (hunk 7 above)
Comment (dcousens) [inline .gitignore]: ? What

File: src/bufferwriter.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
+var bufferutils = require('./bufferutils')
```
Comment (dcousens) [inline src/bufferwriter.js]: I'd rather opt for something like `varstruct` than go this route if possible

File: src/address.js [hunk 11]
```diff
@@ -6,30 +6,54 @@ var types = require('./types')
...
   if (bscript.isScriptHashOutput(scriptPubKey)) return toBase58Check(bscript.compile(scriptPubKey).slice(2, 22), network.scriptHash)
+  if (bscript.isSegWitPubKeyHashOutput(scriptPubKey)) return toBase58Check(bscript.compile(scriptPubKey).slice(2, 22), network.segWitPubKeyHash, 0)
+  if (bscript.isSegWitScriptHashOutput(scriptPubKey)) return toBase58Check(bscript.compile(scriptPubKey).slice(2, 34), network.segWitScriptHash, 0)
 
   throw new Error(bscript.toASM(scriptPubKey) + ' has no matching Address')
 }
 
-function toBase58Check (hash, version) {
-  typeforce(types.tuple(types.Hash160bit, types.UInt8), arguments)
+function toBase58Check (hash, version, segWitVersion) {
```
Comment (dcousens) [inline src/address.js]: It'd be great if we can split out all the address based changes to another PR

**Comment by dcousens** [comment]
Ideally it'd be great to do this in 3 PRs.

`scripts` (https://github.com/bitcoinjs/bitcoinjs-lib/pull/602)
`Transaction`
`TransactionBuilder`
`address`

**Comment by dcousens** [comment]
> Cache hashPrevouts, hashSequence and hashOutputs for all signatures in a transaction (since that was one of the main reasons for changing the signatureHash xD)

Why are we doing such a change in this PR?

**Comment by btcdrak** [comment]
*bump*

**Comment by btcdrak** [comment]
Note for OP, BIP142 was deferred

**Comment by btcdrak** [comment]
The witness program size was extended from 32 to max 40 bytes https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#Witness_program

**Comment by dcousens** [comment]
@btcdrak support for SegWit scripts was already added a small while ago.  
This PR is just serving as a reminder to implement the transaction serialisation logic. 

See https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/script.js#L209-L223 and https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/script.js#L343-L354 and https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/script.js#L382

**Comment by btcdrak** [comment]
Also need to disable uncompressed keys which are disallowed under segwit.

**Comment by dcousens** [comment]
> Also need to disable uncompressed keys which are disallowed under segwit.

Is that in the BIP?

**Comment by btcdrak** [comment]
@dcousens See BIP143 https://github.com/bitcoin/bips/pull/459/files#diff-d6a595a37a5d316054e79e53e9c2382fR128

File: src/transaction.js (hunk 8 above)
Comment (afk11) [inline src/transaction.js]: Not sure if this was cleared up - segwit has a different signature hashing scheme which commits the input amount in the signature, amongst other changes. It has a reusable midstate for all inputs in a transaction which may be cached, so avoids the O(n^2) hashing issue.

**Comment by afk11** [comment]
I might find some time for this coming up, the release including segwit activation is coming up pretty close. There are some breaks in the API, probably passing the sigVersion in places, and the input amount now needs to be known during signing. 

Mostly, sigVersion can be determined in situ, and by including the amount, the full TxOut is now a requirement for signing (previously it was just the scriptPubKey). To address the upcoming BC break, I'd suggest accepting a full txout in the signing API (eg, `hashForSignatureV2`) mostly to avoid cluttering things with extra parameters.

File: src/transaction.js (hunk 9 above)
Comment (afk11) [inline src/transaction.js]: The zero marker makes it obvious to old clients that something is wrong (zero-inputs), and I don't think it will change. Flags may change - it's interpreted as a bit-field where the first indicates segwit support, but setting anything else will intentionally result in an error. https://github.com/bitcoin/bitcoin/blob/2f71490d21796594ca6f55e375558944de9db5a0/src/primitives/transaction.h#L320

**Comment by dcousens** [comment]
>  and the input amount now needs to be known during signing.

That's the first I've seen of that,  but again,  my attention has been diverted?  Any references?

**Comment by dcousens** [comment]
@afk11 do you have a medium with which we can collaborate on that effort?

**Comment by afk11** [comment]
> That's the first I've seen of that, but again, my attention has been diverted? Any references?

I think the 'motivation' is scenarios like a Trezor being fed the wrong UTXO amount, since it blindly signs. If we commit the amount in the sig-hash, full nodes would reject a signature where the wrong amount was used. 

https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#specification 

I'm starting to think for a signing API, the requirements (bar keypairs) are the `TxOut`, `redeemScript`, `witnessScript`, and `hashType`. Looking at it further, I'd love to remove the prevOutScript from addInput, and consolidate it into a txout parameter that replaces `amount` in `TransactionBuilder.sign` as seen in this PR. 

### Inferences

I notice for P2SH the library infers prevOutScript based on the redeemScript. It could probably do this based with segwit scripts too, although I prefer requiring the full TxOut from users always, and only optionally taking a redeemScript / witnessScript. 

Core doesn't infer things, it works from the scriptPubKey and decides what it needs next. (Solver in script/standard.cpp) While checking p2sh?/witness?/bare-scriptPubKey? it determines "is it a standard type, and what sighash sigVersion do I use?". P2SH / P2WSH are treated specially, and involve checking elsewhere to see that the hashes match and another call to Solver. 

Probably not a major issue, but I like there being a chain of hashes indicating which `witnessScript` to use (the reverse of infering the scriptPubKey) because then your wallet works based strictly on what scriptPubKey's do you have the required data for. 

> @afk11 do you have a medium with which we can collaborate on that effort?

Core's slack could work, otherwise happy to use email / XMPP at me$(at)$(IRL name)$(dot)io

**Comment by dcousens** [comment]
> I'd love to remove the prevOutScript from addInput, and consolidate it into a txout parameter that replaces amount in TransactionBuilder.sign as seen in this PR.

Interestingly, I'd prefer to move more to `addInput` and keep `sign` as minimal as possible

**Comment by afk11** [comment]
Ah ok :) I suppose the main reason would be to keep signing atomic - given a transaction, signing only takes so much `(nIn, PrivateKeyInterface key, TransactionOutputInterface txOut, ScriptInterface redeemScript = null, ScriptInterface witnessScript = null, sigHashType = SigHashInterface::ALL)`

Whether we pass all the arguments at signing, or all the arguments at addInput probably doesn't matter, but at least keep it pure, since I don't see any other method to set the prevOutScript (except for directly modifying internal values). I guess my disagreement stems from (i) it being split across two functions and, (ii) we're saving state anticipating it will only be used in that later function, so why not provide it only when necessary?

Also believe the scriptPubKey should be the thing you check a redeemScript against to test suitability when signing, rather than inferring a scriptPubKey from an arbitrary redeemScript and signing with it. I really like how Core work out what to sign with given only a scriptPubKey and it's key/script store: https://github.com/bitcoin/bitcoin/blob/master/src/script/sign.cpp#L139
</opinion>

**Comment by dcousens** [comment]
As discussed on slack, we should set up an example somewhere that idiomatically uses 
P2PKH, P2SH, P2WPKH and P2WSH in succession, with all the necessary pieces of information.
Obviously the `TransactionBuilder` API isn't ready for the task,  but from that point we could play with the API prospectively, seeing what fits nicely and what doesn't.

**Comment by dcousens** [comment]
Updated milestone to `3.1.0`

File: src/transaction.js (hunk 5 above)
Comment (afk11) [inline src/transaction.js]: It was mentioned yesterday, but worth documenting: a param or separate serialization methods are required because you *should* be able to convert a segwit tx into the older version. The other way around (old to new) doesn't make sense, but is still technically valid to witness-enabled software.

File: src/address.js (hunk 11 above)
Comment (afk11) [inline src/address.js]: They shelved that proposal until later - pending feedback/proposals from wallet developers IIRC

**Comment by dcousens** [comment]
Closing in favour of the newer PRs.
Captured some of the checklist with https://github.com/bitcoinjs/bitcoinjs-lib/issues/686 and https://github.com/bitcoinjs/bitcoinjs-lib/issues/685
